src/main/java/hudson/plugins/groovy/WithGroovyStep.java

```diff
@@ -0,0 +1,244 @@
+                    if (_jdk.getName().equals(step.jdk)) {
+                        jdk = _jdk;
+                        break;
+                    }
+                }
+                if (jdk == null) {
+                    throw new AbortException("no such JDK installation " + step.jdk);
+                }
+                jdk = jdk.forNode(getContext().get(Node.class), getContext().get(TaskListener.class));
+                jdk = jdk.forEnvironment(getContext().get(EnvVars.class));
+                String home = jdk.getHome();
+                env.put("PATH+JDK", tmp.child(home).child("bin").getRemote());
+            }
+            if (step.input != null) {
+                try (OutputStream os = tmp.child("input.ser").write(); ObjectOutputStream oos = new ObjectOutputStream(os)) {
+                    oos.writeObject(step.input);
+                }
+                tmp.child("Pipeline.groovy").copyFrom(WithGroovyStep.class.getResource("Pipeline.groovy"));
+                env.put("CLASSPATH+GROOVY", tmp.getRemote());
+            }
+            getContext().newBodyInvoker().
+                withContext(EnvironmentExpander.constant(env)).
+                withCallback(new Callback(tmp)).
+                start();
+            return false;
+        }
+
+        /** Locates {@code groovy-all.jar} on a given node. */
+        private static class FindGroovyAllJAR extends MasterToSlaveCallable<String, IOException> {
+
+            static String runIn(VirtualChannel channel) throws IOException, InterruptedException {
+                // Without the preloadJar call, we would not get back a JAR path, but just a directory containing only groovy/lang/Writable.class.
+                if (channel instanceof Channel) {
+                    ((Channel) channel).preloadJar(WithGroovyStep.class.getClassLoader(), /* arbitrary, but has no other deps to load from agent JVM */ Writable.class);
+                }
+                return channel.call(new FindGroovyAllJAR());
+            }
+
+            @Override
+            public String call() throws IOException {
+                return Which.jarFile(Writable.class).getAbsolutePath();
+            }
+
+        }
+
+    }
+
+    private static class Callback extends BodyExecutionCallback {
+
+        private final String tmp;
+
+        Callback(FilePath tmp) {
+            this.tmp = tmp.getRemote();
+        }
+
+        @Override
+        public void onSuccess(StepContext context, Object result) {
+            try {
+                FilePath ser = context.get(FilePath.class).child(tmp).child("output.ser");
+                if (ser.exists()) {
+                    try (InputStream is = ser.read(); ObjectInputStream ois = new ObjectInputStreamEx(is, GroovyObject.class.getClassLoader(), ClassFilter.DEFAULT)) {
+                        context.onSuccess(ois.readObject());
+                        return;
+                    }
+                }
+            } catch (Exception x) {
```

Comment:
Would it make sense here to catch `Throwable` rather than exception to catch particularly nasty deserialization failures?